core/src/main.cpp

```diff
@@ -847,6 +849,36 @@ auto load_oidc_configuration(const json& _config, json& _oi_config, json& _endpo
 	return true;
 } // load_oidc_configuration
 
+auto load_user_mapping_plugin(const json& _user_mapping_config) -> bool
+{
+	const auto& map_lib{_user_mapping_config["plugin_path"].get_ref<const std::string&>()};
+	const auto config_string{to_string(_user_mapping_config["configuration"])};
+
+	// Load shared lib
+	logging::trace("{}: Loading shared library at [{}]", __func__, map_lib);
+	boost::dll::shared_library user_map_lib{map_lib};
+	irods::http::globals::set_user_mapping_lib(user_map_lib);
+
+	// See if symbols matching interface exist
+	if (!(user_map_lib.has("user_mapper_init") && user_map_lib.has("user_mapper_match") &&
+	      user_map_lib.has("user_mapper_close") && user_map_lib.has("user_mapper_free")))
+	{
+		logging::error("{}: Could not find all required symbols for library [{}]", __func__, map_lib);
+		return false;
+	}
```

Comment:
Still untested currently. I'll try it after addressing the other comments.

Reply:
Did some manual testing with @korydraughn, following is as short list of what was found:

Separate Issues (i.e. not to be handled in this PR):
+ Running HTTP API with config that doesn't exist results in:
  `Error: [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - unexpected end of input; expected '[', '{', or a literal`
  Make missing file error more clear.
  + Create an issue based off this.
+ An iRODS access token is given as long as there is a match, even if the iRODS user does not exist.
  + Ensure users exist before giving out a token. Note, this is likely only for client mode.

Discovered Issues:
+ Add period to 'could not find symbols' log when loading an invalid plugin library.
+ Bad plugin path throws exception.
  + Can leave as is, description is detailed enough.
+ Currently, update is only run on match, meaning invalid mapping files do not cause the server to stop.
  + Run update on init for local_file to fail fast.
+ No validator needed for local_file. If the JSON file can be parsed, we can use it.
  + If there's feedback requesting this, it can be added later.

Up for discussion issue:
+ Making a file invalid after a valid update causes match to fail. Is this the behavior we want?
  + Consider using previous config. If someone gets it wrong, should still match.

---

@trel do comment on anything that catches your eye, especially the up for discussion issue.

@korydraughn feel free to comment on anything you believe I have missed.

Reply:
sounds good - yes, please make new issues for the ones mentioned.

if we can detect an invalid file (various reasons it could be invalid, I suppose).... then we should have some kind of linter or configtest that can fail before we 'load' it and make it live... and it should complain loudly in the logs... and continue to use the valid current config.

Reply:
can we ... get a validator/schema for the local_file?   seems easy and good?

Reply:
> if we can detect an invalid file (various reasons it could be invalid, I suppose).... ... and it should complain loudly in the logs... and continue to use the valid current config.

Gotcha, sounds like a plan.

>... then we should have some kind of linter or configtest that can fail before we 'load' it and make it live... ...

We never make the 'failed' item live; we always keep a valid state. More specifically, we currently fail either on opening the file or parsing JSON. It's just that we also do not allow for matches to happen with an invalid file.

> can we ... get a validator/schema for the local_file? seems easy and good?

@korydraughn also suggested this. I don't think there is much of a schema to write for the local_file plugin. Anything that is valid JSON is mapped to either the OpenID Access Token or ID Token. If we change our mapping file to have the top-level be an array vs. an object, then I think we would need a validator.

We also saw yesterday, that while you can write multiple of the same 'attribute' in a JSON file, it is invalid. However, it will still be parsed and used.
A validator may be useful for this particular case, but I'm not sure if that's something that should run in the plugin vs. a separate executable the user runs themselves.

Reply:
okay - let's make 'write a schema' a new issue and keep moving here.   thanks.    please link new issue here and then resolve.

Reply:
Issue created: #339

Reply:
Addressed in-scope issues.